src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -200,10 +298,103 @@ Score(TreeType& queryNode, TreeType& referenceNode)
+           kernelIsGaussian)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    // Calculate alpha.
+    const double depthAlpha = CalculateAlpha(referenceNode.Stat().Depth());
+    double alpha = depthAlpha + queryStat.AccumAlpha();
+    if (alpha > 1)
+      alpha = 1;
+
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    arma::vec means = arma::zeros(queryNode.NumDescendants());
+    size_t m;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    for (size_t i = 0; i < queryNode.NumDescendants(); ++i)
+    {
+      const size_t queryIndex = queryNode.Descendant(i);
+      sample.clear();
+      m = initialSampleSize;
+      while (m > 0)
+      {
+        const size_t oldSize = sample.size();
+        const size_t newSize = oldSize + m;
+
+        // Don't use probabilistic estimation if this is gonna take a close
+        // amount of computation to the exact calculation.
+        if (newSize >= mcBreakCoef * numDesc)
+        {
+          useMonteCarloPredictions = false;
+          break;
+        }
+
+        // Increase the sample size.
+        sample.resize(newSize);
+        for (size_t i = 0; i < m; ++i)
+        {
+          // Sample and evaluate random points from the reference node.
+          const size_t randomPoint = math::RandInt(0, numDesc);
+          sample(oldSize + i) =
+              EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+        }
+        meanSample = arma::mean(sample);
+        const double stddev = arma::stddev(sample);
+        const double mThreshBase =
+            z * stddev * (1 + relError) / (relError * meanSample);
+        const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+        if (sample.size() < mThresh)
+          m = mThresh - sample.size();
+        else
+          m = 0;
+      }
+
+      if (useMonteCarloPredictions)
+        means(i) = meanSample;
+      else
+        break;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Use Monte Carlo estimation.
+      queryStat.AccumAlpha() = 0;
+      score = DBL_MAX;
+      for (size_t i = 0; i < queryNode.NumDescendants(); ++i)
+        densities(queryNode.Descendant(i)) += numDesc * means(i);
+    }
+    else
+    {
+      // Reclaim not used alpha for Monte Carlo.
+      queryStat.AccumAlpha() += CalculateAlpha(referenceNode.Stat().Depth());
+      // Recurse.
+      score = minDistance;
+    }
   }
   else
   {
     score = minDistance;
+
+    // Reclaim not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      queryStat.AccumAlpha() += CalculateAlpha(referenceNode.Stat().Depth());
```

Comment:
In the code you've written, it looks to me like you are only adding to `AccumAlpha()` when Monte Carlo approximation is not used.  I think this is correct, if the query node (or the query point in the case of single tree) is then evaluated exhaustively with the descendant points of the reference node.  But, the reference node may be pruned by a non-Monte Carlo step (i.e., the difference between minimum and maximum kernel values may be small enough that a "regular" prune can be done).  If that happens, then `AccumAlpha()` will need to be adjusted, as even a regular prune will affect the error that can be used.

Do you think you could write something up (not necessarily in the code, anywhere is fine) detailing the strategy you are using for reclaiming error?  It might be nice to know that we are implementing the right thing by first writing it down, along with either some reasoning for why the result still satisfies the approximation guarantees (perhaps a formal proof would be useful).  That would probably make it a lot easier for me to review too. :+1: